src/ir/properties.h

```diff
@@ -68,6 +69,31 @@ inline bool isNamedControlFlow(Expression* curr) {
   return false;
 }
 
+inline bool isConstantExpression(const Expression* curr) {
+  return curr->is<Const>() || curr->is<RefNull>() || curr->is<RefFunc>();
+}
+
+inline bool mayThrow(Expression* curr) {
+  if (curr->is<Throw>() || curr->is<Rethrow>() || curr->is<Call>() ||
+      curr->is<CallIndirect>()) {
+    return true;
+  }
+  for (auto* child : ChildIterator(curr)) {
+    // Skip inner try bodies. Exceptions thrown within inner tries will be
+    // caught by corresponding inner catches.
+    if (auto* tryy = child->dynCast<Try>()) {
+      if (mayThrow(tryy->catchBody)) {
```

Comment:
It may be risky to use recursion here. Using a walk would use a stack which is safer.

However, can we use an EffectAnalyzer for this? Perhaps "canThrow" is something we should track there.